R/new_matrixfit.R

```diff
@@ -0,0 +1,460 @@
+      res[, 1] <- -xx * exp(-par[1] * xx) * (par[3] + (1 - par[3]) * exp(-par[2] * xx))
+      res[, 2] <- -exp(-par[1] * xx) * (1 - par[3]) * xx * exp(-par[2] * xx)
+      res[, 3] <- exp(-par[1] * xx) * (1 - exp(-par[2] * xx))
+      return(res)
+    },
+    initial_guess = function (corr, parlist, t1, t2) {
+      par = numeric(3)
+
+      ## the ground state energy
+      par[1] <- log(corr[self$reference_time + 1] / corr[self$reference_time + 2])
+      ## the deltaE
+      par[2] <- log(corr[self$reference_time + 1] / corr[self$reference_time + 2]) - par[1]
+      par[2] <- 1.0
+      ## the amplitude
+      par[3] <- 1.0
+
+      return (par)
+    },
+    post_process_par = function (par) {
+      ## The energy and energy difference parameter enters the model only in its
+      ## absolute value, therefore it can become negative. We need to fix that
+      ## here.
+      par[1] <- abs(par[1])
+      par[2] <- abs(par[2])
+      return (par)
+    },
+    reference_time = NA
+  )
+)
+
+#' Model for the n particle correlator fit
+#'
+#' @description
+#' n particle correlator with thermal pollution term(s)
+NParticleModel <- R6Class(
+  'NParticleModel',
+  inherit = MatrixModel,
+  public = list(
+    initialize = function (time_extent, parind, sign_vec, ov_sign_vec) {
+      super$initialize(time_extent, parind, sign_vec, ov_sign_vec)
+      self$sm <- SingleModel$new(time_extent, parind, sign_vec, ov_sign_vec)
+     },
+    prediction = function (par, x, ...) {
+      n <- length(par) / 2
+      y <- 0
+      par_aux <- numeric(2)
+      for (i in 1:n) {
+        par_aux[1] <- par[2*i - 1]
+        par_aux[2] <- par[2*i]
+        corr_part <- self$sm$prediction(par_aux, x, ...)
+        y <- y + corr_part
+      }
+      return(y)
+    },
+    prediction_jacobian = function (par, x, ...) {
+      n <- length(par) / 2
+      par_aux <- numeric(2)
+      par_aux[1] <- par[1]
+      par_aux[2] <- par[2]
+      y <- self$sm$prediction_jacobian(par_aux, x, ...)
+      if (n > 1) {
+        for (i in 2:n) {
+          par_aux[1] <- par[2*i - 1]
+          par_aux[2] <- par[2*i]
+          jacobian_part_aux <- self$sm$prediction_jacobian(par_aux, x, ...)
+          y <- cbind(y, jacobian_part_aux)
+        }
+      }
+      return(y)
+    },
+    sm = NA
+  )
+)
+
+#' @export
+new_matrixfit <- function(cf,
+                          t1, t2,
+                          parlist,
+                          sym.vec,
+                          neg.vec,
+                          useCov = FALSE,
+                          model = "single",
+                          boot.fit = TRUE,
+                          fit.method = "optim",
+                          autoproceed = FALSE,
+                          par.guess,
+                          every,
+                          ...
+                          ) {
+  stopifnot(inherits(cf, 'cf_meta'))
+  stopifnot(inherits(cf, 'cf_boot'))
+  
+  if (model == 'pc') {
```

Comment:
I guess there should be a check for the number of observables in the cf too, right? Otherwise the logic for the 'pc' model below does not make sense.

Reply:
Thank you for your quick reply - I hope that the new checks are okay.